control-userlayer/src/main/java/org/oskari/control/userlayer/CreateUserLayerHandler.java

```diff
@@ -1,222 +1,251 @@
         }
+    }
 
-        if (tempFileBaseName != null) {
-            // rename tempfile if we have an existing basename (shapefiles need to have same basename)
-            final String newName = tempFileBaseName + "." + file.getExtension();
-            File renameLocation = new File(newName);
-            boolean success = newFile.renameTo(renameLocation);
-            if (!success) {
-                //TODO: if shapefile contains more than one layer -> warn user that zip should contain only one layer (shp + dbf + shx)
-                log.warn("Rename failed in temp directory", newFile.getName(), " ->  ", newName);
-            } else {
-                // update path
-                file.setSavedTo(renameLocation.getPath());
-                log.debug("File renamed ", newFile.getName(), " ->  ", newName);
+    private List<File> unZip(FileItem zipFile) throws ActionException {
+        try (InputStream in = zipFile.getInputStream();
+                ZipInputStream zis = new ZipInputStream(in)) {
+            List<File> files = new ArrayList<>();
+            int filesInZip = 0;
+            // Name all the files we find from zip with the same pattern
+            // baseName.{ext} where {ext} is taken from the original filename
+            // Use a 24 letter random UUID as baseName
+            String baseName = UUID.randomUUID().toString().substring(0, 24);
+            ZipEntry ze;
+            while ((ze = zis.getNextEntry()) != null) {
+                filesInZip++;
+                if (filesInZip > MAX_FILES_IN_ZIP) {
+                    // safeguard against infinite loop, userlayers shouldn't have this many files in any case
+                    break;
+                }
+                if (ze.isDirectory()) {
+                    continue;
+                }
+                String ext = getFileExt(ze.getName());
+                if (ext == null) {
+                    continue;
+                }
+                String name = baseName + '.' + ext;
+                File file = writeToTempFile(zis, name);
+                files.add(file);
             }
+            return files;
+        } catch (IOException e) {
+            throw new ActionException("Failed to unzip file", e);
         }
-        return file;
     }
 
-    @Override
-    public void handleAction(ActionParameters params) throws ActionException {
-
-        // stop here if user isn't logged in
-        params.requireLoggedInUser();
-
-        final String target_epsg = params.getHttpParam(PARAM_EPSG_KEY, "EPSG:3067");
-        final String source_epsg = params.getHttpParam(PARAM_SOURCE_EPSG_KEY, defaultSourceEpsg);
-        try {
-
-            // Only 1st file item is handled
-            RawUpLoadItem loadItem = getZipFiles(params);
-
-            // Checks file size
-
-            File file = unZip(loadItem.getFileitem());
+    private String getFileExt(String name) {
+        int i = name.lastIndexOf('.');
+        if (i < 0 || i > name.length() - 1) {
```

Comment:
This could be just i == -1?

Reply:
The first part yes, but the or is there to avoid files named "foo."

Reply:
True, I guess that would return an empty string instead of the null we now get on such a filename? Maybe i +1 == name.length() or name.endsWith(".") would be clearer? With "foo." isn't the i=3 and name.length() - 1 = 3 so the check wouldn't work properly now? Should check this in the morning, it's way too late for thinking this through :D

Reply:
Yea it the comparison should definitely be equal to not greater than for this to work, hehe.